integreat_cms/cms/forms/translations/machine_translation_form.py

```diff
@@ -0,0 +1,112 @@
+import logging
+
+from django import forms
+from django.utils.translation import gettext_lazy as _
+
+from ...utils.translation_utils import mt_to_lang_is_permitted
+
+logger = logging.getLogger(__name__)
+
+
+class MachineTranslationForm(forms.Form):
+    """
+    Form for selecting target languages for machine translation of a content object.
+    """
+
+    automatic_translation = forms.BooleanField(
+        widget=forms.CheckboxInput(),
+        required=False,
+        label=_("Automatic translation"),
+        help_text=_(
+            "Tick if updating this content should automatically refresh or create its translations."
+        ),
+    )
+
+    translations_to_update = forms.ModelMultipleChoiceField(
+        widget=forms.CheckboxSelectMultiple(),
+        queryset=None,
+        required=False,
+        label=_("Update existing translations:"),
+    )
+
+    translations_to_create = forms.ModelMultipleChoiceField(
+        widget=forms.CheckboxSelectMultiple(),
+        queryset=None,
+        required=False,
+        label=_("Create new translations:"),
+    )
+
+    class Meta:
+        fields = [
+            "automatic_translation",
+            "translations_to_update",
+            "translations_to_create",
+        ]
+
+    def __init__(self, instance, region, language, **kwargs):
+        r"""
+        Initialize MT translation form
+
+        :param \**kwargs: The supplied keyword arguments
+        :type \**kwargs: dict
+        """
+        super().__init__(**kwargs)
+
+        self.initial["automatic_translation"] = False
+
+        translation_targets = [
+            language
+            for language in region.language_tree_nodes.filter(parent=language.id)
+            if language.mt_provider and mt_to_lang_is_permitted(region, language.slug)
+        ]
+
+        # completely hide options if no target languages exist
+        if not translation_targets:
+            del self.fields["automatic_translation"]
+            return
+
+        to_update, to_create = [], []
+        for target in translation_targets:
+            target_type = (
+                to_update
+                if instance and instance.get_translation(target.slug)
+                else to_create
+            )
+            target_type.append((target.id, target.translated_name))
+
+        self.fields["translations_to_update"].choices = to_update
+        self.fields["translations_to_create"].choices = to_create
+
+        self.fields[
+            "translations_to_update"
+        ].queryset = region.language_tree_nodes.filter(
+            id__in=[language[0] for language in to_update]
+        )
+        self.fields[
+            "translations_to_create"
+        ].queryset = region.language_tree_nodes.filter(
+            id__in=[language[0] for language in to_create]
+        )
+
+        self.initial["translations_to_update"] = [language[0] for language in to_update]
```

Comment:
Good to know. The only reason I was using `queryset` here at all is because I saw in your implementation [here](https://github.com/digitalfabrik/integreat-cms/blob/c6a29e4ef0c73ec29641f468b0aabb7b60825ad2/integreat_cms/cms/templates/translations/translations_management.html#L45) that it was possible to check for the existence of choices in a template this way, and it seems with just `choices` that's unfortunately not possible. Is there a reason both `choices` and `queryset` in the [corresponding form](https://github.com/digitalfabrik/integreat-cms/blob/c6a29e4ef0c73ec29641f468b0aabb7b60825ad2/integreat_cms/cms/forms/translations/translations_management_form.py#L81)?

Reply:
I think `queryset` is always required for a model (multiple) choice form, and in this case I explicitly used choices to allow the grouping by mt provider (there is no possibility to set [optgroups](https://github.com/digitalfabrik/integreat-cms/blob/c6a29e4ef0c73ec29641f468b0aabb7b60825ad2/integreat_cms/cms/templates/translations/languages_input.html#L2) just by setting the queryset I think.

In this case, we could also have used a normal choice field instead of a model choice field to save a few sql queries, but in the end, a model choice field is probably the cleaner way to do it, so I'd leave it that way.

Reply:
Yes, it originally was a non-model `MultipleChoiceForm`. But that overly complicated checking if the list of choices was empty inthe template.